jobsdb/readonly_jobsdb.go

```diff
@@ -22,7 +24,7 @@ import (
 ReadonlyJobsDB interface contains public methods to access JobsDB data
 */
 type ReadonlyJobsDB interface {
-	HavePendingJobs(customValFilters []string, count int, parameterFilters []ParameterFilterT) bool
+	HavePendingJobs(ctx context.Context, customValFilters []string, count int, parameterFilters []ParameterFilterT) bool
```

Comment:
The current solution encapsulates error handling behaviour as implementation details of the`ReadonlyJobsDB#HavePendingJobs` method. The method treats an `error` as a `true` result, instead of letting the caller handle the error appropriately.
If we have any need in the future to use `ReadonlyJobsDB#HavePendingJobs` for some other purpose and want to be able to interpret an error in a different way, then we should simplify it already.

Based on the above, I propose to change the method signatures of all methods inside `readonly_jobsdb.go` (both private and public) which now accept a `ctx` to return the error instead of handling it on their own, so that the error can bubble up to the caller.

1. Change the following method signatures:
```
func (jd *ReadonlyHandleT) HavePendingJobs(ctx context.Context, customValFilters []string, count int, parameterFilters []ParameterFilterT) bool, error
func (jd *ReadonlyHandleT) haveUnprocessedJobs(ctx context.Context, customValFilters []string, parameterFilters []ParameterFilterT) bool, error
func (jd *ReadonlyHandleT) getUnprocessedJobsDSCount(ctx context.Context, ds dataSetT, customValFilters []string, parameterFilters []ParameterFilterT) int64, error
func (jd *ReadonlyHandleT) haveNonSucceededJobs(ctx context.Context, customValFilters []string, parameterFilters []ParameterFilterT) bool, error
func (jd *ReadonlyHandleT) haveProcessedJobs(ctx context.Context, stateFilter []string, customValFilters []string, parameterFilters []ParameterFilterT) bool, error
func (jd *ReadonlyHandleT) haveProcessedJobs(ctx context.Context, stateFilter []string, customValFilters []string, parameterFilters []ParameterFilterT) bool, error
```
2. Modify method implementations so that they return errors when they happen, instead of handling them internally.
3. Catch the error at the `gateway` and treat it as `pending=true`

It should really be the responsibility of the caller, i.e. the `gateway` to decide how to interpret an error and in this specific case, the `gateway` can decide that whenever an error occurs it should assume that pending events do exist.

Reply:
Changed HavePendingJobs. Since that is the only function gateway calls. We can change other functions later.